lib/routes/utils/oauth.js

```diff
@@ -0,0 +1,61 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+const ScopeSet = require('fxa-shared').oauth.scopes;
+
+// right now we only care about notifications for the following scopes
+// if not a match, then we don't notify
+const NOTIFICATION_SCOPES = ScopeSet.fromArray(['https://identity.mozilla.com/apps/oldsync']);
+
+module.exports = {
+  newTokenNotification: async function newTokenNotification (db, oauthdb, mailer, devices, request, grant) {
+    const clientId = request.payload.client_id;
+    const scopeSet = ScopeSet.fromString(grant.scope);
+    const credentials = request.auth && request.auth.credentials || {};
+
+    if (! scopeSet.intersects(NOTIFICATION_SCOPES)) {
+      // right now we only care about notifications for the `oldsync` scope
+      // if not a match, then we don't do any notifications
+      return;
+    }
+
+    if (! credentials.uid) {
+      // this can be removed once issue #3000 has been resolved
+      const tokenVerify = await oauthdb.checkAccessToken({
+        token: grant.access_token
+      });
+      // some grant flows won't have the uid in `credentials`
+      credentials.uid = tokenVerify.user;
+    }
+
+    if (! credentials.refreshTokenId) {
+      // provide a refreshToken for the device creation below
+      credentials.refreshTokenId = grant.refresh_token;
+    }
+
+    // we set tokenVerified because the granted scope is part of NOTIFICATION_SCOPES
+    credentials.tokenVerified = true;
+    credentials.client = await oauthdb.getClientInfo(clientId);
+
+    // The following upsert gets no `deviceInfo`.
+    // However, `credentials` provides a `client` which generates the name for the device
```

Comment:
nit: it took me a little bit to understand what this comment was trying to say; it might be clearer worded along the lines of "However, `credentials.client` lets it generate a default name for the device".